inference: allows conditional object to propagate constraint multiple times

aviatesk opened the pull request:
Currently we always `widenconditional` conditional var state, which
makes it unable to propagate constraints from conditional object
multiple times:
```julia
@test Base.return_types((Union{Nothing,Int},)) do a
    b = a === nothing
    c = b ? 0 : a # c::Int
    d = !b ? a : 0 # d::Int ideally, but Union{Int,Nothing}
    c, d
 end == Any[Tuple{Int,Int}] # fail
```

This PR keeps conditional var state when the update is came from a
conditional branching, and allows a conditional object to propagate
constraint multiple times as far as the subject of condition doesn't
change. AFAIU this is safe because the update from conditional
branching doesn't affect the validity of the condition itself.

aviatesk commented on base/compiler/typelattice.jl:
```diff
@@ -321,7 +322,8 @@ function stupdate!(state::Nothing, changes::StateUpdate)
         # remove any Conditional for this Slot from the vtable
-        for i = 1:length(newst)
+        # (unless this change is came from the conditional)
+        !changes.conditional && for i = 1:length(newst)
```
I use this strange code style in order to minimize the diff, but I'm welcome to use the basic `if` block style if you want.

vtjnash replied:
Yeah, could you change this to standard style? I'll usually hide whitespace after I verify formatting, to examine the functional diff

aviatesk replied:
Sure thing. Change to the standard style, and also rebased against the current master.

aviatesk committed:
inference: allows conditional object to propagate constraint multiple times

Currently we always `widenconditional` conditional var state, which
makes us unable to propagate constraints from conditional object
multiple times:
```julia
@test Base.return_types((Union{Nothing,Int},)) do a
    b = a === nothing
    c = b ? 0 : a # c::Int
    d = !b ? a : 0 # d::Int ideally, but Union{Int,Nothing}
    c, d
 end == Any[Tuple{Int,Int}] # fail
```

This PR keeps conditional var state when the update is came from a
conditional branching, and allows a conditional object to propagate
constraint multiple times as far as the subject of condition doesn't
change. AFAIU this is safe because the update from conditional
branching doesn't change the condition itself.

aviatesk commented:
@vtjnash bump ?
